ofctrl/fgraphSwitch.go

```diff
@@ -54,13 +54,6 @@ func (self *OFSwitch) initFgraph() error {
 	normalLookup.outputType = "normal"
 	normalLookup.portNo = openflow15.P_NORMAL
 	self.normalLookup = normalLookup
-
-	// Clear all existing flood lists
-	groupMod := openflow15.NewGroupMod()
-	groupMod.GroupId = openflow15.OFPG_ALL
-	groupMod.Command = openflow15.OFPGC_DELETE
-	groupMod.Type = openflow15.GT_ALL
-	return self.Send(groupMod)
```

Comment:
You're correct. And there is a comment in the Antrea Agent code about this: https://github.com/antrea-io/antrea/blob/454df5dd383eac93aedab4df6f5b0e4b478e46db/pkg/agent/agent.go#L465-L470

I don't know if things have changed with Openflow 1.5. If this code is not needed any more, the comment in the Antrea Agent code should be updated as well.

Reply:
IIRC, this is quite easy to test. You just need to kill the antrea-agent process / container and let it restart. The idea is that the Agent should restart, _without OVS restarting_. So you don't want to delete the antrea-agent Pod. If there is any error during Agent initialization after the restart, then we either need this code or more complex logic in the Antrea Agent.

Reply:
I didn't notice this usage. Then I would revert to the original implementation.

Reply:
I didn't see errors in Antrea logs in agent restart case. But I think a better way is to remove the existing groups before replaying groups in the reconnections in Antrea Agent code. So I add this part in another change in antrea repo.

Reply:
then it's strange we didn't have group cleanup logic but there seems no issue with it. Let me confirm how the groups are cleaned up first.

Reply:
I can confirm the groups were indeed cleaned up once antrea-agent restarts, then the question is how it's triggered.

Reply:
Maybe the groups are always deleted in case of reconnection, and there is nothing new here? As @wenyingd pointed out, I was originally mistaken and this code only deletes groups with type ALL (and not all groups, which would require ANY).

But it opens up the following questions:
1) is it the same for flows? When I added the cookie mechanism to delete stale flows 2+ years ago, I believe it was to solve an actual problem
2) what about meters? When we added meters for packet-in rate limiting, we had to add code to the Antrea agent to delete meters in case of restart to avoid the `OFPMMFC_METER_EXISTS` error: https://github.com/antrea-io/antrea/pull/2215#issuecomment-850016931. Are meters treated differently from flows / groups, that seems strange.
3) we actually _don't_ want flows / groups to be deleted on antrea-agent restart, to minimize traffic disruption as it can take some time to restore all flows. Can this behavior be modified by changing the fail-mode of the OVS bridge from standalone to secure?

Reply:
@wenyingd I'm reading the Openflow spec and I think that the group type is ignored for a `OFPGC_DELETE` operation. So setting it to ALL should not matter. This code should be deleting all groups of all types.

Did you test without this code, with a restart of the antrea-agent container?

Reply:
> Maybe the groups are always deleted in case of reconnection, and there is nothing new here?

I checked the existing code in antrea, and I didn't find related logic to delete the existing groups in re-connect case. Before the flexible pipeine logic, I don't remember we have the logic to replay groups. Actually, I doubt we don't test the case to replay groups in a reconnect scenario.

> But it opens up the following questions:
> 
> 1. is it the same for flows? When I added the cookie mechanism to delete stale flows 2+ years ago, I believe it was to solve an actual problem

Flow is different. There is a special field "cookie" in every flow, and we use several bits in the cookie to identify round number. Every time antrea is re-connected to OVS, we update the round number, then replay the flows with new cookie ID (including the new round number), and then we delete the flows with filter of the last round. Since the flows are override with the new cookie ID if they are replayed, the deletion operation only removes the flows not needed. But the cookie is not existing in group modification message.

> 2. what about meters? When we added meters for packet-in rate limiting, we had to add code to the Antrea agent to delete meters in case of restart to avoid the `OFPMMFC_METER_EXISTS` error: [Packet-in rate limiting with OF Meter antrea#2215 (comment)](https://github.com/antrea-io/antrea/pull/2215#issuecomment-850016931). Are meters treated differently from flows / groups, that seems strange.

I  didn't see meter replay operations in antrea existing code.

> 3. we actually _don't_ want flows / groups to be deleted on antrea-agent restart, to minimize traffic disruption as it can take some time to restore all flows. Can this behavior be modified by changing the fail-mode of the OVS bridge from standalone to secure?

In my understanding, standalone mode means the openflow entries are deleted once the OF controller is disconnected, and secure mode means the existing flows are deleted afterOVS is re-connected to OF Controller. The difference is if the existing flows are kept during the period when OVS is disconnected from Controller. I am not sure if the logic is also suitable on the groups and meters.  But anyway, the flows are empty when antrea agent is reconnected to OVS, and this is a different behavior from the current logic.

Reply:
> @wenyingd I'm reading the Openflow spec and I think that the group type is ignored for a `OFPGC_DELETE` operation. So setting it to ALL should not matter. This code should be deleting all groups of all types.
> 
> Did you test without this code, with a restart of the antrea-agent container?

Let me try.

Reply:
Even if the operation to delete existing groups is needed, I don't think is a good practice to leave this behavior in ofnet. Shall we move it to antrea? Otherwise, we have to always remember that ofnet has this pre-configuration. @antoninbas @tnqn

Reply:
> > @wenyingd I'm reading the Openflow spec and I think that the group type is ignored for a `OFPGC_DELETE` operation. So setting it to ALL should not matter. This code should be deleting all groups of all types.
> > Did you test without this code, with a restart of the antrea-agent container?
> 
> Let me try.

@antoninbas You are correct, if I remove this code ( deleteing all groups in agent restart only case), I would got this error when replaying groups: